File: R/hammingDist.R
hammingDist <- function(G1, G2, allMistakesOne = TRUE)
    # hammingDist(G1,G2)
    #
    # Computes Hamming Distance between DAGs G1 and G2 with SHD(->,<-) = 1 if allMistakesOne == TRUE
    #
    # INPUT:  G1, G2     adjacency graph containing only zeros and ones: (i,j)=1 means edge from X_i to X_j.
    #
    # OUTPUT: hammingDis Hamming Distance between G1 and G2
    #
    # Copyright (c) 2012-2013  Jonas Peters [peters@stat.math.ethz.ch]
    # All rights reserved.  See the file COPYING for license terms.
{
    if(allMistakesOne)
    {
        Gtmp <- (G1 + G2) %% 2
        Gtmp <- Gtmp + t(Gtmp)
        nrReversals <- sum(Gtmp == 2) / 2
        nrInclDel <- sum(Gtmp == 1) / 2
        hammingDis <- nrReversals + nrInclDel
    } else
    {
        hammingDis <- sum(abs(G1 - G2))
        # correction: dist(-,.) = 1, not 2
        hammingDis <- hammingDis - 0.5 * sum(G1 * t(G1) * (1 - G2) * t(1 - G2) + G2 * t(G2) * (1 - G1) * t(1 - G1))
    }

    return(hammingDis)
}


read_testcase <- function(filepath)
{
    lines <- readLines(filepath, warn = FALSE)

    G1 <- list()
    G2 <- list()
    mode <- NULL

    for(line in lines)
    {
        line <- trimws(line)

        if(line == "")
        {
            next
        }

        if(line == "Matrix1:")
        {
            mode <- "matrix1"
            next
        }

        if(line == "Matrix2:")
        {
            mode <- "matrix2"
            next
        }

        if(mode == "matrix1")
        {
            G1[[length(G1) + 1]] <- as.integer(strsplit(line, " +")[[1]])
        } else if(mode == "matrix2")
        {
            G2[[length(G2) + 1]] <- as.integer(strsplit(line, " +")[[1]])
        }
    }

    G1 <- do.call(rbind, G1)
    G2 <- do.call(rbind, G2)

    return(list(G1 = G1, G2 = G2))
}

#llm
script_dir <- getwd()
project_dir <- dirname(script_dir)

testcase_dir <- file.path(project_dir, "tests", "hammingDist", "Testcase")
output_dir <- file.path(project_dir, "tests", "hammingDist", "R_outputs")

if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
}

output_file <- file.path(output_dir, "all_results.txt")
out <- file(output_file, open = "w")

for(i in 1:100)
{
    filepath <- file.path(testcase_dir, paste0(i, ".txt"))
    values <- read_testcase(filepath)

    G1 <- values$G1
    G2 <- values$G2

    result <- hammingDist(G1, G2)

    writeLines(paste0("Testcase ", i), out)

    writeLines("Matrix1:", out)
    for(r in 1:nrow(G1)) {
        writeLines(paste(G1[r, ], collapse = " "), out)
    }

    writeLines("Matrix2:", out)
    for(r in 1:nrow(G2)) {
        writeLines(paste(G2[r, ], collapse = " "), out)
    }

    writeLines("Result:", out)
    writeLines(as.character(result), out)

    writeLines("", out)
}

close(out)

File: py_impl/hammingDist.py
import os
import numpy as np


def hamming_dist(G1, G2, all_mistakes_one=True):
    G1 = np.array(G1, dtype=int)
    G2 = np.array(G2, dtype=int)

    if all_mistakes_one:
        Gtmp = (G1 + G2) % 2
        Gtmp = Gtmp + Gtmp.T
        nr_reversals = np.sum(Gtmp == 2) / 2
        nr_incl_del = np.sum(Gtmp == 1) / 2
        hamming_dis = nr_reversals + nr_incl_del
    else:
        hamming_dis = np.sum(np.abs(G1 - G2))
        hamming_dis = hamming_dis - 0.5 * np.sum(
            G1 * G1.T * (1 - G2) * (1 - G2).T +
            G2 * G2.T * (1 - G1) * (1 - G1).T
        )

    return hamming_dis


def read_testcase(filepath):
    G1 = []
    G2 = []
    mode = None

    with open(filepath, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()

            if line == "":
                continue

            if line == "Matrix1:":
                mode = "matrix1"
                continue

            if line == "Matrix2:":
                mode = "matrix2"
                continue

            if mode == "matrix1":
                G1.append([int(x) for x in line.split()])
            elif mode == "matrix2":
                G2.append([int(x) for x in line.split()])

    return np.array(G1, dtype=int), np.array(G2, dtype=int)


#llm

script_dir = os.getcwd()
project_dir = os.path.dirname(script_dir)

testcase_dir = os.path.join(project_dir, "tests", "hammingDist", "Testcase")
output_dir = os.path.join(project_dir, "tests", "hammingDist", "Py_outputs")

if not os.path.exists(output_dir):
    os.mkdir(output_dir)

output_file = os.path.join(output_dir, "all_results.txt")

print("Reading testcases from:", testcase_dir)
print("Saving output to:", output_file)

with open(output_file, "w", encoding="utf-8") as f:
    for i in range(1, 101):
        filepath = os.path.join(testcase_dir, f"{i}.txt")
        G1, G2 = read_testcase(filepath)

        result = hamming_dist(G1, G2)

        f.write(f"Testcase {i}\n")

        f.write("Matrix1:\n")
        for row in G1:
            f.write(" ".join(map(str, row)) + "\n")

        f.write("Matrix2:\n")
        for row in G2:
            f.write(" ".join(map(str, row)) + "\n")

        f.write("Result:\n")
        f.write(str(result) + "\n")

        f.write("\n")

print("Done. Output saved to:", output_file)